### DISCUSSION TITLE
Add support for liking songs

### DISCUSSION THREAD

**Discussion by mainrs** [librespot-org/librespot#625] [Feature Requests]
Is it currently possible to like the song that is being played?

I couldn't find any relevant API for it.

**Comment by ashthespy** [comment]
Don't think anyone has actively looked into it. 
[`librespot-org/spotify-analyze`](https://github.com/librespot-org/spotify-analyze) is your best friend when it comes to these things..

**Comment by devgianlu** [comment]
@ashthespy Pretty much all that's there doesn't work anymore. All the communication except for metrics, audio key (and some other) are done through HTTPS.

BTW, the new API provides a feedback on whether the song is liked/disliked:
```
metadata {
    key: "radio.thumb"
    value: "up"
}
```
I once also saw the endpoint to make the choice, but I'm not seeing it now.

**Comment by willstott101** [comment]
Hm, there are two UI elements at work here aren't there? A favourite/saved song and a liked/disliked radio song. AFAIK they're two different things. One adds to your collection, the other feeds the recommendation engine and adds to a playlist?

Though I have no idea if that playlist is likely to be a separate client action or not.

**Comment by mainrs** [comment]
I am looking for ages to find a program/library that allows it. I wanted to build a simple button that would then just like the current playing song that I could attach to my desk. 

@willstott101 I was referring to the `heart` one, the one that saves it to your collection and influences your recommendations too.

**Comment by devgianlu** [comment]
>  A favourite/saved song and a liked/disliked radio song. AFAIK they're two different things.

There should indeed be two options, but when you click the heart button, it'll return `radio.thumb` as true. I also wasn't able to play something that would bring up the two thumb icons.

**Comment by ashthespy** [comment]
From what I recall, `v1/me/tracks/` is the end point for all the heart/like stuff.
Every track is checked against  `v1/me/tracks/contains?ids=` which responds with a boolean.

**Comment by mainrs** [comment]
> 
> 
> > A favourite/saved song and a liked/disliked radio song. AFAIK they're two different things.
> 
> There should indeed be two options, but when you click the heart button, it'll return `radio.thumb` as true. I also wasn't able to play something that would bring up the two thumb icons.

You didn't find an endpoint to actually set `radio.thumb` to true, did you?

**Comment by devgianlu** [comment]
> You didn't find an endpoint to actually set radio.thumb to true, did you?

It used to be in the context info, but I can look it up again. Hopefully it is via Hermes.

**Comment by devgianlu** [comment]
I found a working endpoint for dailymixes:

    hm://dailymix/v3/mixes/like/song/0z89qn3koTrXqvu4ds0xi7/7krbSH3rd8lhIZvuzTV3Bl

Where `0z89qn3koTrXqvu4ds0xi7` should be the dailymix and `7krbSH3rd8lhIZvuzTV3Bl` the track.

**Comment by ashthespy** [comment]
For the like/unlike (aka thumbs up/down)
```
hm://feedback/v1/feedback/dislike?uri=
hm://feedback/v1/feedback/?uri=
```
For the liked songs, I only found some references via the webgate:
```
# To heart a song:
PUT https://spclient.wg.spotify.com/v1/me/tracks?ids=07HwlohZWn3ssYatDQLyPg
# To un-heart a song
`DELETE https://spclient.wg.spotify.com/v1/me/tracks?ids=07HwlohZWn3ssYatDQLyPg
# To check if a song is hearted/unhearted 
GET https://spclient.wg.spotify.com/v1/me/tracks/contains?ids=07HwlohZWn3ssYatDQLyPg&market=from_token
```
The webSDK makes the same calls, but to `https://api.spotify.com/` instead.

**Comment by mainrs** [comment]
Haha, thanks guys, really appreciate your help!

**Comment by devgianlu** [comment]
These are the official endpoints the new API sends for the Release Radar playlist:
```
context_metadata {
      key: "dislike-endpoint"
      value: "new-jams/feedback/v1/RELEASE_RADAR/DISLIKE?uri=<track-uri>&reason=<reason>&version=26027882&context=spotify:user:spotify:playlist:37i9dQZEVXbdSfbIwIVcBT"
}

context_metadata {
      key: "like-endpoint"
      value: "new-jams/feedback/v1/RELEASE_RADAR/LIKE?uri=<track-uri>&reason=<reason>&version=26027882&context=spotify:user:spotify:playlist:37i9dQZEVXbdSfbIwIVcBT"
}
```

**Comment by p4vook** [comment]
Any updates on this?
